### PR TITLE
Add CUDA graph capture to the core Trainer

### PR DESCRIPTION
**Pull request by xmfan**
Stacked PRs:
 * __->__#3559

--- --- ---

#### Known limitations
- Pipeline parallelism primarily needs a different capture boundary rather than different model semantics: capture each stage forward/backward action with its static microbatch state while leaving the Python schedule and P2P handoff eager, instead of capturing the current combined forward+loss+backward callable.
- Any captured path that synchronizes with the CPU is unsupported. This includes blocking AllToAll expert parallelism; CUDA graph EP currently requires nonblocking HybridEP or MinimalAsyncEP.
- FSDP reshard_after_forward=always is unsupported. Its all-gather work uses hidden FSDP streams, and capture can encounter a dependency on an event produced by uncaptured work, resulting in CUDA_ERROR_STREAM_CAPTURE_ISOLATION. The default and never policies work.

#### No warmup needed
Numerics already match without warmup iteration.
1st iteration: record
2nd+ iteration: replay

#### Copy-in
Some tensors need to be copied each iteration due to the dataloader's sequence packing. For DSV3 16B these are:
  - Input token IDs
  - Labels
  - global_valid_tokens scalar
  - Position IDs
  - FlexAttention BlockMask tensors:
      - block counts and indices
      - full-block metadata
      - backward/query ordering metadata
      - tensors captured by the packed-document mask function

#### Test Plan
- Llama 3 debug model, 1 GPU, 10 deterministic steps: eager and CUDA graph TensorBoard loss and grad norm were exactly equal at every step.
- DeepSeek v3 debug model, 1 GPU, balanced MoE routing, 10 deterministic steps: eager and CUDA graph TensorBoard loss and grad norm were exactly equal at every step. This GB300 SM103 build required a temporary bmm test shim because native torch._grouped_mm fails to launch; the shim is not included in this commit.
- 8x H100, 10 deterministic steps: compared raw TensorBoard float32 loss and grad norm values with CUDA graphs disabled and enabled. All 60 scalar pairs matched bitwise (maximum absolute difference 0):
  - Llama 3 70B, FSDP=4 + TP=2
  - DeepSeek v3 16B HybridEP, FSDP=4 + EP=2 + TP=2
  - DeepSeek v3 16B HybridEP, FSDP=8 + EP=8, Full AC/SAC/No AC
  - DeepSeek v3 16B MinimalAsyncEP, FSDP=8 + EP=8, Full AC

Profile w/ CG: https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/xmfan_9ba1b77a-f72b-4e7d-b7ad-716941df298c_rank0_trace.json.gz

can see the cg launch from cpu thread
image

#### Next steps

There still remains some sync (loss accumulation) after the optimizer to eliminate in the trainer, to be removed soon.
PP uses a different fwd bwd entrypoint.

Disabled cudagraph configs:
  - Pipeline-parallel configs
  - Standard host-synchronizing EP
  - Varlen attention
  - Flux/Kimi/Muse multimodal configs

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -683,28 +683,93 @@ def forward_backward_step(
...
+        if not hasattr(self, "_cg_wrapper"):
+            model = self.model_parts[0]
+            train_context = self.train_context
+            loss_fn = self.loss_fn
+            full_dtensor_enabled = self.config.parallelism.full_dtensor
+            disable_loss_parallel = self.config.parallelism.disable_loss_parallel
+
+            def _step(inp, lab, gvt, *extra_flat):
+                with train_context():
+                    pred = model(inp, **extra_inputs, **extra_kwargs)
```
Comment (weifengpy) [inline torchtitan/trainer.py]: is this cudagraphing eager parallellisms as well, say FSDP2?

Comment (xmfan) [reply torchtitan/trainer.py]: yeah it does, everything that model.forward + loss.backward() runs (single cudagraph for whole model, not piecewise cudagraphs). I had issues with FSDP + AC, which is why the examples I tried here are using AC=none

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -129,6 +134,15 @@ def __post_init__(self):
...
 
+            if (
+                self.training.enable_cuda_graphs
+                and self.parallelism.pipeline_parallel_degree > 1
+            ):
+                raise ValueError(
+                    "training.enable_cuda_graphs does not support pipeline "
+                    "parallelism yet."
+                )
+
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: question 1: what it takes to support PP + cudagraph?

question 2:
Let's test composability with other features, including SAC, compile, etc. We should test both features.py and models.py with cudagraph enabled and see where it breaks (we don't have to fix all of them in this PR)

Comment (xmfan) [reply torchtitan/trainer.py]: im working on those locally, let me ping u when it's ready for review

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -319,7 +319,7 @@ def __call__(
         self,
         pred: torch.Tensor,
         labels: torch.Tensor,
-        global_valid_tokens: float | None = None,
+        global_valid_tokens: float | torch.Tensor | None = None,
```
Comment (xmfan) [inline torchtitan/components/loss.py]: cudagraphs need this data to live on gpu, so we need a tensor. we could make it always a tensor (on cpu w/o cg)

File: torchtitan/trainer.py [hunk 4]
```diff
@@ -165,6 +183,40 @@ def __post_init__(self):
...
+                )
+                if not isinstance(
+                    dispatcher_config, MinimalAsyncEPTokenDispatcher.Config
+                )
+                and not (
+                    isinstance(dispatcher_config, HybridEPTokenDispatcher.Config)
+                    and dispatcher_config.non_blocking_capacity_factor is not None
+                )
+            }
+            if unsupported_dispatchers:
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: instead of collecting all unsupported, we could early raise on first violation

File: torchtitan/distributed/cudagraph.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+
+@dataclass(frozen=True)
+class _BlockMaskInputSpec:
+    num_leaves: int
+    context: tuple[Any, ...]
+
+
+class CUDAGraphInputSpec:
+    """Flatten structured inputs while exposing tensors stored in ``BlockMask``."""
```
Comment (tianyu-l) [inline torchtitan/distributed/cudagraph.py:40]: We are going to move mask creation into model code, after which do we no longer need this class? If so please add a TODO for @jinsooihm to remove

File: torchtitan/trainer.py [hunk 6]
```diff
@@ -229,6 +281,7 @@ def maybe_log(self) -> None:
     device: torch.device
     gc_handler: utils.GarbageCollection
     train_context: dist_utils.SpmdContext
+    _fwd_bwd_runner: _ForwardBackward
```
Comment (tianyu-l) [inline torchtitan/trainer.py]:
```suggestion
    fwd_bwd_fn: ForwardBackwardFn
```
underscore seems the right thing, but just to be consistent with other existing namings

File: torchtitan/trainer.py
Comment (tianyu-l) [inline torchtitan/trainer.py:781]: There're too many fwd bwd functions in this class. I would propose this high-level idea:
- only having one for non-pp, and one for pp -- we don't need a central `forward_backward_step` that calls if-else within
- set `fwd_bwd_fn` to one of (1) non-pp (2) pp (3) cudagraph wrapped non-pp
- call `fwd_bwd_fn` instead of `forward_backward_step` or `pp_forward_backward_step`

Please let me know if anything doesn't make sense.

File: torchtitan/trainer.py [hunk 7]
```diff
@@ -810,10 +876,81 @@ def pp_forward_backward_step(
             return torch.sum(torch.stack(losses)).to(self.device)
         return torch.tensor([-1.0], device=self.device)
 
+    def _wrap_with_cuda_graph(
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: seems this entire function can go into distributed/cudagraph.py as it's pretty stateless

File: torchtitan/trainer.py
Comment (xmfan) [reply torchtitan/trainer.py:781]: You can't cudagraph these things currently in `forward_backward_step`: 
- inputs/extra_kwargs prep + attention-mask update
- ntokens_seen logging
- CP sharding and DTensor/SPMD input preparation
- raw dictionary/list handling

Which is why only the "body" is cudagraphed. 

> set fwd_bwd_fn to one of (1) non-pp (2) pp (3) cudagraph wrapped non-pp

This is a bit weird, we'd then have (4) cudagraph wrapped pp. cudagraph is just a feature to apply on top of any of the existing branches.

pp/non-pp have different pre-processing and post-processing, we'd be hiding the `forward_backward_step` branching into preprocess/postprocess branches:
  - Preprocessing: same hook, but non-PP prepares one microbatch while PP prepares lists of microbatches for the pipeline schedule.
  - Execution/postprocessing:
      - Non-PP directly runs model, loss, and backward().
      - PP delegates to the schedule, aggregates losses on the last stage, and returns a sentinel elsewhere.
      - CUDA graphs retain non-PP semantics but manage static inputs and clone the replayed loss.

File: torchtitan/trainer.py [hunk 8]
```diff
@@ -752,8 +801,17 @@ def forward_backward_step(
         inputs, labels, extra_kwargs = self.post_dataloading_process(input_dict, labels)
```
Comment (tianyu-l) [inline torchtitan/trainer.py:802]: The current code looks clear enough, but I'm not sure how it'd work with future changes like https://github.com/pytorch/torchtitan/pull/4116. Could you help review?

I also wonder what if we put the current `post_dataloading_process` inside model code, including attention mask generation, and CP sharding, would it break cudagraph?

File: torchtitan/config/configs.py [hunk 9]
```diff
@@ -52,6 +52,18 @@ class TrainingConfig:
     Whether to apply CPU offloading of parameters, gradients, and optimizer states in FSDP
     """
 
+    enable_cuda_graphs: bool = False
```
Comment (tianyu-l) [inline torchtitan/config/configs.py]: Could you enable cudagraph in CI wherever we can? IIUC multimodal and PP (and some EP backend) don't work, but others should?

File: torchtitan/distributed/cudagraph.py (hunk 5 above)
Comment (fegin) [reply torchtitan/distributed/cudagraph.py:40]: I don't think mask creation is CUDAGraph-able. So this will need anyway -- creating mask outside of CUDAGraph region and pass it into the CUDAGraph region, which means that you will need this class. 

@xmfan is this correct?

File: torchtitan/trainer.py (hunk 8 above)
Comment (fegin) [reply torchtitan/trainer.py:802]: > I also wonder what if we put the current post_dataloading_process inside model code, including attention mask generation, and CP sharding, would it break cudagraph?

It will break.

**Comment by fegin** [comment]
I suggest that we land this PR to enable CUDAGraph. We need some discussion on post-processing refactor to make it work with CUDAGraph.

cc., @jinsooihm

File: tests/integration_tests/run_tests.py [hunk 10]
```diff
@@ -143,6 +144,10 @@ def run_single_test(
         cmd = f'TORCH_TRACE="{output_dir}/{test_name}/compile_trace" ' + cmd
 
         cmd += " " + dump_folder_arg
+        if test_flavor.enable_cuda_graphs and gpu_arch_type == "cuda":
+            # Tyro requires global options to precede union subcommands such as
+            # activation-checkpoint:selective.
+            cmd += " --training.enable_cuda_graphs"
```
Comment (tianyu-l) [inline tests/integration_tests/run_tests.py]: I think it's better if we directly use `--training.enable_cuda_graphs` in the tests, and make `wrap_with_cuda_graph` no-op when it's not cuda. This will avoid the introduction of `test_flavor.enable_cuda_graphs` and make it robust against other hardware like rocm.

File: torchtitan/config/configs.py (hunk 9 above)
Comment (tianyu-l) [reply torchtitan/config/configs.py]: After seeing the list in CI where we enabled cudagraph, I feel we should keep it by default on and only disable on the ones that doesn't support cudagraph. Then it'd be quite visible what are things we need to work on.

Maybe we should rename this flag to `disable_cuda_graphs: bool = False`.

**Review by tianyu-l — APPROVED** [review]
looks great, thank you!

**Comment by xmfan** [comment]
Disabled cudagraph configs:
  - Pipeline-parallel configs
  - Standard host-synchronizing EP
  - Varlen attention
  - Flux/Kimi/Muse multimodal configs

**Comment by tianyu-l** [comment]
Do you know what happens to varlen attention? @liangel-02 used to work to enable full cudagraph for inference (prefill + decode), so I feel it should be cudagraphable for training.

File: .github/workflows/integration_test_8gpu_features.yaml [hunk 11]
```diff
@@ -139,10 +139,13 @@ jobs:
...
         echo "Checking Qwen3 MoE FSDP+TP+EP loss parity against reference"
+        # qwen3_moe_debug uses EP=1, where CUDA graphs are supported. This test
+        # overrides EP to 4 without changing the AllToAll token dispatcher,
+        # whose CPU synchronization is incompatible with CUDA graph capture.
         python3 scripts/loss_compare.py . . \
           --baseline-module=qwen3 --baseline-config=qwen3_moe_debug \
-          --baseline-options="--parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 4" \
-          --test-options="--parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 4" \
+          --baseline-options="--parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 4 --training.disable_cuda_graphs" \
+          --test-options="--parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 4 --training.disable_cuda_graphs" \
```
Comment (xmfan) [inline .github/workflows/integration_test_8gpu_features.yaml:148]: @tianyu-l lmk if you have suggestions on how to handle this cleanly. `disable_cuda_graphs` lives on each model config, but some default configs are trivially valid w/ cg e.g. EP=1. A common pattern is to override the config with ep degree, in which case cg may or may not be supported.

**Comment by xmfan** [comment]
> Do you know what happens to varlen attention? @liangel-02 used to work to enable full cudagraph for inference (prefill + decode), so I feel it should be cudagraphable for training.

The VarlenMetadata needs to be migrated to holding tensors and copied in each iteration, there's also a sync to remove:
https://github.com/pytorch/torchtitan/blob/2d8b2ecf37905f04794faf2c5a576eb1f34ae999/torchtitan/models/common/attention.py#L672

File: .github/workflows/integration_test_8gpu_features.yaml (hunk 11 above)
Comment (tianyu-l) [reply .github/workflows/integration_test_8gpu_features.yaml:148]: It looks fine to add this flag everywhere EP (with non-cudagraphable backend) is used. Can revisit later.
